File: 프로그래머스 고득점 kit/해시/완주하지 못한 선수.py
# 1회차 순수하게 품
# def solution(participant, completion):
#     cp = {}

#     for p in participant:
#         if cp.get(p) is None:
#             cp[p] = 1
#         else:
#             cp[p] += 1

#     for p in completion:
#         if cp[p] > 0:
#             cp[p] -= 1

#     for person, cnt in cp.items():
#         if cnt > 0:
#             return person

# 다른 사람 풀이 Counter 객체 사용
from collections import Counter


def solution(participant, completion):
    answer = Counter(participant) - Counter(completion)
    return list(answer.keys())[0]


print(solution(["leo", "kiki", "eden"],	["eden", "kiki"]) == "leo")
print(solution(["marina", "josipa", "nikola", "vinko", "filipa"],
               ["josipa", "filipa", "marina", "nikola"]) == "vinko")
print(solution(["mislav", "stanko", "mislav", "ana"],
               ["stanko", "ana", "mislav"]) == "mislav")

File: 프로그래머스 고득점 kit/해시/완주하지 못한 선수.js
function solution(participant, completion) {
  const hashedParticipant = {}

  participant.forEach((p) => {
    if (!!hashedParticipant[p]) {
      hashedParticipant[p] += 1
    } else {
      hashedParticipant[p] = 1
    }
  })

  for (const c of completion) {
    hashedParticipant[c] -= 1
    if (hashedParticipant[c] === 0) {
      delete hashedParticipant[c]
    }
  }

  return Object.entries(hashedParticipant)[0][0]
}

console.log(solution(['leo', 'kiki', 'eden'], ['eden', 'kiki']) === 'leo')
console.log(solution(['marina', 'josipa', 'nikola', 'vinko', 'filipa'], ['josipa', 'filipa', 'marina', 'nikola']) === 'vinko')
console.log(solution(['mislav', 'stanko', 'mislav', 'ana'], ['stanko', 'ana', 'mislav']) === 'mislav')
